MonoGame.Framework/Graphics/Vertices/IndexBuffer.cs

```diff
@@ -115,10 +115,26 @@ internal protected override void GraphicsDeviceResetting()
         protected void SetDataInternal<T>(int offsetInBytes, T[] data, int startIndex, int elementCount, SetDataOptions options) where T : struct
         {
             if (data == null)
-                throw new ArgumentNullException("data is null");
+                throw new ArgumentNullException("data");
             if (data.Length < (startIndex + elementCount))
                 throw new InvalidOperationException("The array specified in the data parameter is not the correct size for the amount of data requested.");
 
+            var type = typeof (T);
+            switch (IndexElementSize)
+            {
+                case IndexElementSize.SixteenBits:
+                    if (type != typeof(short) || type != typeof(ushort))
+                    {
+                        throw new ArrayTypeMismatchException("Invalid type using when short or ushort need to be used.");
+                    }
+                    break;
+                case IndexElementSize.ThirtyTwoBits:
+                    if (type != typeof(int) || type != typeof(uint))
+                    {
+                        throw new ArrayTypeMismatchException("Invalid type using when int or uint need to be used.");
+                    }
+                    break;
+            }

```

Comment:
So the failure above shows at least shows something is wrong there.

We need to add a test where the `short[]` array is copied into a `byte[]` for instance and then `SetData`/`GetData` is called.  I suspect `IndexBuffer` doesn't care what `T` is as long as it can be copied to the buffer properly.